google/cloud/firestore_v1/base_document.py

```diff
@@ -503,6 +504,48 @@ def _to_protobuf(self) -> Optional[Document]:
         return _helpers.document_snapshot_to_protobuf(self)
 
 
+class DocumentSnapshotList(list):
+    """A list of DocumentSnapshot.
+
+    This is a subclass of the built-in list. A new property `explain_metrics`
+    is added to return the query profile results.
+
+    Args:
+        reference (:class:`~google.cloud.firestore_v1.document.DocumentReference`):
+            A document reference corresponding to the document that contains
+            the data in this snapshot.
+        docs (list[DocumentSnapshot]):
+            The list of DocumentSnapshot.
+        explain_options
+            (Optional[:class:`~google.cloud.firestore_v1.query_profile.ExplainOptions`]):
+            Options to enable query profiling for this query. When set,
+            explain_metrics will be available on the returned generator.
+        explain_metrics (Optional[query_profile.ExplainMetrics]):
+            Query profile results.
+    """
+
+    def __init__(
+        self,
+        docs: list,
+        explain_options: Optional[query_profile.ExplainOptions] = None,
+        explain_metrics: Optional[query_profile.ExplainMetrics] = None,
+    ):
+        super().__init__(docs)
+        self._explain_options = explain_options
+        self._explain_metrics = explain_metrics
+
+    @property
+    def explain_options(self):
+        return self._explain_options
+
+    @property
+    def explain_metrics(self):
+        if self._explain_options is None:
+            raise query_profile.QueryExplainError("explain_options not set on query.")
+        else:
+            return self._explain_metrics
```

Comment:
I agree this is repeating code, but also I feel like using mixin for one property might be a bit of overkill? I think you have a good point that we are raising a lot of exceptions with exactly the same error message. Would it make sense to create a few subclasses of QueryExplainError for each of the cases, so at least we have a central place to update the error messages?

Reply:
Yeah, I think you're right, Mixins probably only make sense if the shared code starts to expand, so maybe just keep it in mind if it seems it may be useful at some point. Probably overkill for now

One other thought I mentioned in a different thread though: DocumentSnapshotList could hold an instance of StreamGenerator, and delegate this call to the inner stream instance, which would be another way to reduce duplication

Reply:
Great idea! I will add it along with the other list related changes